### DISCUSSION TITLE
A separate API for creating a root from a device

### DISCUSSION THREAD

**Discussion by iwoplaza** [software-mansion/TypeGPU#484] [Ideas]
Currently, a root can be created with the `tgpu.init` method. When passing a device, it will wrap it instead of requesting a new one.

```ts
const root = await tgpu.init();
// or
const root = await tgpu.init({ device });
```

Note that both calls are asynchronous, even though the second call could be synchronous. It would be weird and inconsistant to return either a Promise of something, or that thing depending on the passed in parameters.

If we keep the existing-device root creation asynchronous though, we can encounter some troubling limitations:

### Limitation 1: Having to wrap everything in an immediately-invoked function
In cases like `react-native-wgpu`, we get the device in a callback:

```ts
useWebGPU(({ context, device, presentationFormat }) => {
  (async () => {
    const root = await tgpu.init({ device });

    // Creating resources...
  })();
});
```

### Limitation 2: Hard to create resources in a constructor

Passing devices throughout the app is a common pattern in WebGPU apps. When incorporating TypeGPU, we have to create
roots in sync-places (e.g., constructors):

```ts
class GameObject {
  constructor(device: GPUDevice) {
    const root = await tgpu.init({ device }); // <- cannot await here :(
    
    // Creating resources...
  }
}
```

## Proposed solution

A separate API for initializing a root from an existing device that is synchronous:
```ts
const root = await tgpu.init();
// or
const root = await tgpu.init({ ... }); // <- device descriptor, adapter, etc.
// or
const root = tgpu.initFromDevice({ device });
```

**Comment by iwoplaza** [comment]
@mhawryluk @reczkok Thoughts? 🧠

**Comment by mhawryluk** [comment]
agreed, makes total sense

**Comment by reczkok** [comment]
Great idea 💡
